x/concentrated-liquidity/internal/math/tick.go

```diff
@@ -2,46 +2,123 @@ package math
-	// If the tick index is positive, we can use the normal equation to calculate the square root price.
-	// However, if the tick index is negative, since we cannot take negative powers with the sdk,
-	// we need to take one over the original equation in order to make the power positive.
-	if tickIndex.GTE(sdk.ZeroInt()) {
-		sqrtPrice, err = sdk.NewDecWithPrec(10001, 4).Power(tickIndex.Uint64()).ApproxSqrt()
+// TickToSqrtPrice calculates the price at a given tick index based on the provided
+// starting price of 1 at k=0. The price is calculated using a square root function
+// with a coefficient of 9, where the price increases by a factor of 10 for every
+// increment of k.
+//
+// The function takes in two arguments:
+// 	- tickIndex: the tick index to calculate the price for
+// 	- kAtPriceOne: the value of k at which the starting price of 1 is set
+//
+// It returns a sdk.Dec representing the calculated price and an error if any errors
+// occurred during the calculation.
+func TickToPrice(tickIndex, kAtPriceOne sdk.Int) (price sdk.Dec, err error) {
+	if tickIndex.IsZero() {
+		return sdk.OneDec(), nil
+	}
+
+	var kIncrementDistance sdk.Dec
+	// The formula is as follows: k_increment_distance = 9 * 10**(-k_at_price_1)
+	// Due to sdk.Power restrictions, if the resulting power is negative, we take 9 * (1/10**k_at_price_1)
+	if kAtPriceOne.GTE(sdk.ZeroInt()) {
+		kIncrementDistance = sdk.NewDec(9).Mul(sdk.OneDec().Quo(sdk.NewInt(10).ToDec().Power(kAtPriceOne.Uint64())))
 	} else {
-		sqrtPrice, err = sdk.OneDec().Quo(sdk.NewDecWithPrec(10001, 4).Power(tickIndex.Abs().Uint64())).ApproxSqrt()
+		kIncrementDistance = sdk.NewDec(9).Mul(sdk.NewInt(10).ToDec().Power(kAtPriceOne.Abs().Uint64()))
 	}
-	if err != nil {
-		return sdk.Dec{}, err
+
+	// Use floor division to determine how many k increments we have passed
+	kDelta := tickIndex.ToDec().Quo(kIncrementDistance).TruncateInt()
+
+	// Calculate the current k value from the starting k value and the k delta
+	curK := (kAtPriceOne.Add(kDelta))
+
+	var curIncrement sdk.Dec
+	if curK.GTE(sdk.ZeroInt()) {
+		curIncrement = sdk.NewDec(10).Power(curK.Uint64())
+	} else {
+		curIncrement = sdk.NewDec(1).Quo(sdk.NewDec(10).Power(curK.Abs().Uint64()))
 	}
 
-	return sqrtPrice, nil
+	numAdditiveTicks := tickIndex.ToDec().Sub(kDelta.ToDec().Mul(kIncrementDistance))
+
+	if kDelta.GTE(sdk.ZeroInt()) {
+		price = sdk.NewDec(10).Power(kDelta.Uint64()).Add(numAdditiveTicks.Mul(curIncrement))
+	} else {
+		price = sdk.OneDec().Quo(sdk.NewDec(10).Power(kDelta.Abs().Uint64())).Add(numAdditiveTicks.Mul(curIncrement))
+	}
+	return price, nil
 }
 
 // PriceToTick takes a price and returns the corresponding tick index
-func PriceToTick(price sdk.Dec) sdk.Int {
-	tick := osmomath.BigDecFromSDKDec(price).CustomBaseLog(osmomath.NewDecWithPrec(10001, 4))
-	return tick.SDKDec().TruncateInt()
+func PriceToTick(price sdk.Dec, kAtPriceOne sdk.Int) (tickIndex sdk.Int) {
```

Comment:
this named return is redundant

```suggestion
func PriceToTick(price sdk.Dec, kAtPriceOne sdk.Int) sdk.Int {
```